File: cost/iona/orbacus/cpp/poa/TestAdapterActivatorServer.cpp
// **********************************************************************
//
// Copyright (c) 2001
// Object Oriented Concepts, Inc.
// Billerica, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

#include <includes.h>

#include <TestUtil.h>

#include <Test_skel.h>

#include <Test_impl.h>
#include <TestServer_impl.h>

#include <string.h>

#include <fstream>

using namespace CORBA;
using namespace PortableServer;

//
// Sun CC 5.0 can't handle this function
//
#if 0
static POA_ptr
createTestPOA(POA_ptr parent)
{
    POAManager_var mgr = parent -> the_POAManager();

    PolicyList policies(3);
    policies.length(3);
    policies[0] =
        parent -> create_lifespan_policy(PortableServer::PERSISTENT);
    policies[1] =
        parent -> create_id_assignment_policy(PortableServer::USER_ID);
    policies[2] = parent ->
        create_request_processing_policy(PortableServer::USE_SERVANT_MANAGER);

    return parent -> create_POA("poa3", mgr, policies);
}
#endif


class TestAdapterActivator_impl
    : public POA_PortableServer::AdapterActivator,
      public PortableServer::RefCountServantBase
{
    String_var expectedName_;
    bool create_;
    bool invoked_;

public:

    TestAdapterActivator_impl()
        : create_(false), invoked_(false)
    {
    }

    void reset(const char* name, bool create)
    {
        expectedName_ = string_dup(name);
        create_ = create;
        invoked_ = false;
    }

    bool invoked() const
    {
        return invoked_;
    }

    virtual Boolean unknown_adapter(POA_ptr parent, const char* name)
        throw(SystemException)
    {
        TEST(strcmp(name, expectedName_) == 0);
        invoked_ = true;
        if(create_)
        {
            PolicyList policies;
            POAManager_var mgr = parent -> the_POAManager();
            POA_var poa = parent -> create_POA(name, mgr, policies);
            return true;
        }
        else
            return false;
    }
};

void
TestAdapterActivator(ORB_ptr orb, POA_ptr root)
{
    Object_var obj;
    PolicyList policies;
    POA_var poa, parent;
    POAManager_var mgr;
    String_var str;

    POAManager_var rootMgr = root -> the_POAManager();
    TEST(!is_nil(rootMgr));

    TestAdapterActivator_impl* activatorImpl = new TestAdapterActivator_impl();
    PortableServer::ServantBase_var activatorServant = activatorImpl;
    AdapterActivator_var activator = activatorImpl -> _this();
    root -> the_activator(activator);

    //
    // Test: Activator and successful creation
    //
    activatorImpl -> reset("poa1", true);
    poa = root -> find_POA("poa1", true);
    TEST(!is_nil(poa));
    TEST(activatorImpl -> invoked());
    str = poa -> the_name();
    TEST(strcmp(str, "poa1") == 0);
    parent = poa -> the_parent();
    TEST(!is_nil(parent));
    TEST(parent -> _is_equivalent(root));

    //
    // Test: Activator and unsuccessful creation
    //
    activatorImpl -> reset("poa2", false);
    try
    {
        root -> find_POA("poa2", true);
    }
    catch(const POA::AdapterNonExistent&)
    {
        // expected
    }
    TEST(activatorImpl -> invoked());

    //
    // Test: Make sure activator isn't called when POA already exists
    //
    activatorImpl -> reset("poa1", true);
    poa = root -> find_POA("poa1", true);
    TEST(!activatorImpl -> invoked());

    //
    // Test: Disable adapter activator and make sure it isn't invoked
    //
    root -> the_activator(AdapterActivator::_nil());
    activatorImpl -> reset("poa2", false);
    try
    {
        root -> find_POA("poa2", true);
    }
    catch(const POA::AdapterNonExistent&)
    {
        // expected
    }
    TEST(!activatorImpl -> invoked());

    poa -> destroy(true, true);
}

//
// Classes for testing the adapter activator on a remote call.
//
class TestServantActivator_impl :
    public virtual POA_PortableServer::ServantActivator,
    public virtual PortableServer::RefCountServantBase
{
    ORB_var orb_;

public:

    TestServantActivator_impl(ORB_ptr orb)
        : orb_(ORB::_duplicate(orb))
    {
    }

    Servant incarnate(const ObjectId& oid, POA_ptr poa)
        throw(ForwardRequest, SystemException)
    {
        String_var oidString = ObjectId_to_string(oid);

        //
        // If the user is requesting the object "test" then oblige
        //
        Servant servant = 0;
        if(strcmp("test", oidString) == 0)
            servant = new Test_impl(orb_, "test", false);
        if(servant != 0)
        {
            //
            // Verify that POA allows activator to explicitly activate
            // a servant
            //
#if 0
	    // DG: This is an invalid test. The POA spec does not say
	    // this should be allowed.
	    poa -> activate_object_with_id(oid, servant);
#endif
            return servant;
        }

        //
        // Fail
        //
        throw OBJECT_NOT_EXIST();
        return 0; // Some compilers need this
    }

    void etherealize(const ObjectId& oid,
                     POA_ptr /*poa*/,
                     Servant servant,
                     Boolean /*cleanup*/,
                     Boolean remaining)
        throw(SystemException)
    {
        if(!remaining)
        {
            String_var oidString = ObjectId_to_string(oid);

            //
            // If the user is requesting the object "test" then oblige.
            //
            if(strcmp("test", oidString) == 0)
                delete servant;
        }
    }
};

class TestRemoteAdapterActivator_impl
    : public POA_PortableServer::AdapterActivator,
      public PortableServer::RefCountServantBase
{
    ORB_var orb_;
    PortableServer::POA_var root_;
    TestServantActivator_impl* activator_;

public:

    TestRemoteAdapterActivator_impl(ORB_ptr orb, PortableServer::POA_ptr root)
        : orb_(ORB::_duplicate(orb)),
          root_(PortableServer::POA::_duplicate(root))
    {
        activator_ = new TestServantActivator_impl(orb_);
    }

    ~TestRemoteAdapterActivator_impl()
    {
        activator_ -> _remove_ref();
    }

    virtual Boolean unknown_adapter(POA_ptr parent, const char* name)
        throw(SystemException)
    {
        if(strcmp(name, "poa3") == 0)
        {
            // Sun CC bug
            //POA_var poa = createTestPOA(parent);

            PortableServer::POAManager_var mgr = parent -> the_POAManager();

            CORBA::PolicyList policies(3);
            policies.length(3);
            policies[0] =
                parent -> create_lifespan_policy(PortableServer::PERSISTENT);
            policies[1] =
                parent -> create_id_assignment_policy(PortableServer::USER_ID);
            policies[2] = parent -> create_request_processing_policy(
                PortableServer::USE_SERVANT_MANAGER);

	    POA_var poa = parent -> create_POA("poa3", mgr, policies);

            ServantActivator_var activator = activator_ -> _this();
            poa -> set_servant_manager(activator);
            return true;
        }
        else
            return false;
    }
};

int
main(int argc, char* argv[], char*[])
{
    int status = EXIT_SUCCESS;
    ORB_var orb;

    try
    {
        orb = ORB_init(argc, argv);

        POA_var root = TestUtil_GetRootPOA(orb);

        POAManager_var manager = root -> the_POAManager();
        TEST(!is_nil(manager));

        //
        // Run implementation
        //
        manager -> activate();

        TestAdapterActivator(orb, root);

        //
        // First create an object-reference to the test POA. Then
        // destroy the POA so that the adapter activator will activate
        // the POA when necessary.
        //
        // Sun CC bug
        //POA_var poa = createTestPOA(root);
        PortableServer::POAManager_var mgr = root -> the_POAManager();

        CORBA::PolicyList policies(3);
        policies.length(3);
        policies[0] =
            root -> create_lifespan_policy(PortableServer::PERSISTENT);
        policies[1] =
            root -> create_id_assignment_policy(PortableServer::USER_ID);
        policies[2] = root -> create_request_processing_policy(
            PortableServer::USE_SERVANT_MANAGER);

	POA_var poa = root -> create_POA("poa3", mgr, policies);

        ObjectId_var oid = string_to_ObjectId("test");
        Object_var reference1 =
            poa -> create_reference_with_id(oid, "IDL:Test:1.0");
        poa -> destroy(true, true);
        poa = POA::_nil();
        
        //
        // Create server
        //
        TestInfoSeq info;
        info.length(1);
        info[0].obj = Test::_narrow(reference1);
        info[0].except_id = string_dup("");
        TestServer_impl* serverImpl = new TestServer_impl(orb, info);
        TestServer_var server = serverImpl -> _this();
        
        //
        // Save reference
        //
        const char* refFile = "Test.ref";
        ofstream out(refFile);
        if(out.fail())
        {
            OB_ERROR(argv[0] << ": can't open `" << refFile << "'");
            return 1;
        }
        
        String_var ref = orb -> object_to_string(server);
        out << ref << endl;
        out.close();
        
        TestRemoteAdapterActivator_impl* activatorImpl =
	    new TestRemoteAdapterActivator_impl(orb, root);
	PortableServer::ServantBase_var activatorServant = activatorImpl;
        AdapterActivator_var activator = activatorImpl -> _this();
        root -> the_activator(activator);
        
        orb -> run();

        //
        // Clean up
        //
        serverImpl -> _remove_ref();

        unlink(refFile);
    }
    catch(const Exception& ex)
    {
#ifdef HAVE_EXCEPTION_INSERTERS
	OB_ERROR(ex);
#else
	cerr << "Exception: " << ex._rep_id() << endl;
#endif
        status = EXIT_FAILURE;
    }

    if(!is_nil(orb))
    {
	try
	{
	    orb -> destroy();
	}
	catch(const Exception& ex)
	{
#ifdef HAVE_EXCEPTION_INSERTERS
	    OB_ERROR(ex);
#else
	    cerr << "Exception: " << ex._rep_id() << endl;
#endif
	    status = EXIT_FAILURE;
	}
    }
    
    return status;
}

File: cost/legacy/iona/orbacus/java/poa/TestAdapterActivatorServer.java
// **********************************************************************
//
// Copyright (c) 2000
// Object Oriented Concepts, Inc.
// Billerica, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

package test.poa;

import org.omg.CORBA.*;
import org.omg.PortableServer.*;
import org.omg.PortableServer.POAPackage.*;
import org.omg.PortableServer.POAManagerPackage.*;
import java.io.*;

final class TestAdapterActivatorServer extends test.common.TestBase
{
    static POA
    createTestPOA(POA parent)
    {
        POAManager mgr = parent.the_POAManager();

        Policy[] policies = new Policy[3];
        policies[0] = parent.create_lifespan_policy(
            org.omg.PortableServer.LifespanPolicyValue.PERSISTENT);
        policies[1] = parent.create_id_assignment_policy(
            org.omg.PortableServer.IdAssignmentPolicyValue.USER_ID);
        policies[2] = parent.create_request_processing_policy(
            org.omg.PortableServer.RequestProcessingPolicyValue.
            USE_SERVANT_MANAGER);

        try
        {
            return parent.create_POA("poa3", mgr, policies);
        }
        catch(AdapterAlreadyExists ex)
        {
            TEST(false);
        }
        catch(InvalidPolicy ex)
        {
            TEST(false);
        }

        return null;
    }

    final static class TestAdapterActivator_impl extends AdapterActivatorPOA
    {
        private String expectedName_;
        private boolean create_;
        private boolean invoked_;

        TestAdapterActivator_impl()
        {
            create_ = false;
            invoked_ = false;
        }

        void
        reset(String name, boolean create)
        {
            expectedName_ = name;
            create_ = create;
            invoked_ = false;
        }

        boolean
        invoked()
        {
            return invoked_;
        }

        public boolean
        unknown_adapter(POA parent, String name)
        {
            TEST(name.equals(expectedName_));
            invoked_ = true;
            if(create_)
            {
                Policy[] policies = new Policy[0];
                POAManager mgr = parent.the_POAManager();
                try
                {
                    POA poa = parent.create_POA(name, mgr, policies);
                }
                catch(AdapterAlreadyExists ex)
                {
                    TEST(false);
                }
                catch(InvalidPolicy ex)
                {
                    TEST(false);
                }
                return true;
            }
            else
                return false;
        }
    }

    static void
    TestAdapterActivator(ORB orb, POA root)
    {
        org.omg.CORBA.Object obj;
        Policy[] policies;
        POA poa, parent;
        POAManager mgr;
        String str;

        POAManager rootMgr = root.the_POAManager();
        TEST(rootMgr != null);

        //
        // If this function terminates due to an exception this
        // application will crash because this stack based servant will
        // still be registered with the root POA
        //
        TestAdapterActivator_impl activatorImpl =
            new TestAdapterActivator_impl();
        AdapterActivator activator = activatorImpl._this(orb);
        root.the_activator(activator);

        //
        // Test: Activator and successful creation
        //
        activatorImpl.reset("poa1", true);
        poa = null;
        try
        {
            poa = root.find_POA("poa1", true);
        }
        catch(AdapterNonExistent ex)
        {
            TEST(false);
        }
        TEST(poa != null);
        TEST(activatorImpl.invoked());
        str = poa.the_name();
        TEST(str.equals("poa1"));
        parent = poa.the_parent();
        TEST(parent != null);
        TEST(parent._is_equivalent(root));

        //
        // Test: Activator and unsuccessful creation
        //
        activatorImpl.reset("poa2", false);
        try
        {
            root.find_POA("poa2", true);
        }
        catch(AdapterNonExistent ex)
        {
            // expected
        }
        TEST(activatorImpl.invoked());

        //
        // Test: Make sure activator isn't called when POA already exists
        //
        activatorImpl.reset("poa1", true);
        try
        {
            poa = root.find_POA("poa1", true);
        }
        catch(AdapterNonExistent ex)
        {
            TEST(false);
        }
        TEST(!activatorImpl.invoked());

        //
        // Test: Disable adapter activator and make sure it isn't invoked
        //
        root.the_activator(null);
        activatorImpl.reset("poa2", false);
        try
        {
            root.find_POA("poa2", true);
        }
        catch(AdapterNonExistent ex)
        {
            // expected
        }
        TEST(!activatorImpl.invoked());

        poa.destroy(true, true);

        try
        {
            byte[] id = root.servant_to_id(activatorImpl);
            root.deactivate_object(id);
        }
        catch(ServantNotActive ex)
        {
            TEST(false);
        }
        catch(ObjectNotActive ex)
        {
            TEST(false);
        }
        catch(WrongPolicy ex)
        {
            TEST(false);
        }
    }

    //
    // Classes for testing the adapter activator on a remote call.
    //
    final static class TestServantActivator_impl extends ServantActivatorPOA
    {
        private ORB orb_;

        TestServantActivator_impl(ORB orb)
        {
            orb_ = orb;
        }

        public Servant
        incarnate(byte[] oid, POA poa)
            throws ForwardRequest
        {
            String oidString = new String(oid);

            //
            // If the user is requesting the object "test" then oblige
            //
            Servant servant = null;
            if(oidString.equals("test"))
                servant = new Test_impl(orb_, "test", false);
            if(servant != null)
            {
                //
                // Verify that POA allows activator to explicitly activate
                // a servant
                //
                try
                {
                    poa.activate_object_with_id(oid, servant);
                    return servant;
                }
                catch(ObjectAlreadyActive ex)
                {
                    TEST(false);
                }
                catch(ServantAlreadyActive ex)
                {
                    TEST(false);
                }
                catch(WrongPolicy ex)
                {
                    TEST(false);
                }
            }

            //
            // Fail
            //
            throw new OBJECT_NOT_EXIST();
        }

        public void
        etherealize(byte[] oid,
                    POA poa,
                    Servant servant,
                    boolean cleanup,
                    boolean remaining)
        {
            if(!remaining)
            {
                String oidString = new String(oid);

                //
                // If the user is requesting the object "test" then oblige.
                //
                if(oidString.equals("test"))
                    servant = null;
            }
        }
    }

    final static class TestRemoteAdapterActivator_impl
        extends AdapterActivatorPOA
    {
        private ORB orb_;
        private POA root_;
        private TestServantActivator_impl activator_;

        TestRemoteAdapterActivator_impl(ORB orb, POA root)
        {
            orb_ = orb;
            root_ = root;
            activator_ = new TestServantActivator_impl(orb_);
            activator_._this(orb_);
        }

        public boolean
        unknown_adapter(POA parent, String name)
        {
            if(name.equals("poa3"))
            {
                POA poa = createTestPOA(parent);

                ServantActivator activator = activator_._this();
                try
                {
                    poa.set_servant_manager(activator);
                }
                catch(WrongPolicy ex)
                {
                    TEST(false);
                }
                return true;
            }
            else
                return false;
        }
    }

    public static void
    main(String[] args)
    {
        java.util.Properties props = System.getProperties();

        ORB orb = null;

        try
        {
            orb = ORB.init(args, props);

            POA root = TestUtil.GetRootPOA(orb);

            POAManager manager = root.the_POAManager();
            TEST(manager != null);

            //
            // Run implementation
            //
            try
            {
                manager.activate();
            }
            catch(org.omg.PortableServer.POAManagerPackage.AdapterInactive ex)
            {
                TEST(false);
            }

            TestAdapterActivator(orb, root);

            //
            // First create an object-reference to the test POA. Then
            // destroy the POA so that the adapter activator will activate
            // the POA when necessary.
            //
            POA poa = createTestPOA(root);

            byte[] oid = ("test").getBytes();
            org.omg.CORBA.Object reference1 = null;
            reference1 = poa.create_reference_with_id(oid, "IDL:test/poa/Test:1.0");
            poa.destroy(true, true);
            poa = null;
            
            //
            // Create server
            //
            TestInfo[] info = new TestInfo[1];
            info[0] = new TestInfo();
            info[0].obj = TestHelper.narrow(reference1);
            info[0].except_id = "";
            TestServer_impl serverImpl = new TestServer_impl(orb, info);
            TestServer server = serverImpl._this(orb);

            //
            // Save reference
            //
            String refFile = "Test.ref";
            try
            {
                FileOutputStream file = new FileOutputStream(refFile);
                PrintWriter out = new PrintWriter(file);
                out.println(orb.object_to_string(server));
                out.flush();
                file.close();
            }
            catch(IOException ex)
            {
                System.err.println("Can't write to `" +
                                   ex.getMessage() + "'");
                System.exit(1);
            }

            TestRemoteAdapterActivator_impl activatorImpl =
                new TestRemoteAdapterActivator_impl(orb, root);
            AdapterActivator activator = activatorImpl._this(orb);
            root.the_activator(activator);
            
            orb.run();

            File file = new File(refFile);
            file.delete();
        }
        catch(SystemException ex)
        {
            ex.printStackTrace();
            System.exit(1);
        }

        if(orb != null)
        {
            try
            {
                orb.destroy();
            }
            catch(SystemException ex)
            {
                ex.printStackTrace();
                System.exit(1);
            }
        }

        System.exit(0);
    }
}
